File: bin/ssip/seqalign/build_cpx_alignment.cpp
//###################################################################################################
//# build_complex_alignment
//#
//# This is a perl script used to build the protein complex MSA from two monomer MSAs via the STRING 
//# protein links database.
//# Input:  <-- protein.links.txt
//#         <-- pdb_lig.msa
//#         <-- pdb_rec.msa
//# Output: --> protein_cpx_alignment.txt
//#
//# Xiaoqiang Huang
//# xiaoqiah@umich.edu
//###################################################################################################

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX_SPE_PAIR  1000
#define MAX_PRO_PAIR  1000

#define MAX_DEF_LENG  100
#define MAX_LIN_LENG  10000
#define MAX_SEQ_LENG  10000
#define MAX_NAM_LENG  100


char *linkfile = NULL;
char *msa1file = NULL;
char *msa2file = NULL;
char *cpxoutfile = NULL;


typedef struct _Comm{
  int spe;
  int count;
  char content[MAX_PRO_PAIR][MAX_DEF_LENG];
}Comm;


int make_new_files(){
  char cmd[1000];
  sprintf(cmd, "cp -rf %s msa1\n", msa1file);
  system(cmd);
  sprintf(cmd, "cp -rf %s msa2\n", msa2file);
  system(cmd);
  return 0;
}


int sort_msa(){
  printf("Sorting ligand and receptor MSA files ... \n");
  system("sort -k2 msa1 > sorted_msa1\n");
  system("sort -k2 msa2 > sorted_msa2\n");
  printf("Done.\n");
  return 0;
}

// find common organisms of the two MSAs and build a pseudolink 
// between two proteins for each organism
int find_common_species_and_build_links(){
  Comm* comms=(Comm*)malloc(sizeof(Comm)*MAX_SPE_PAIR);
  printf("Find common species of two MSAs and build links between proteins ... \n");
  // set a maximum common specie number: 1000
  for(int i = 0; i < MAX_SPE_PAIR; i++){
    comms[i].count = 0;
    // for each common specie, set the maximum protein pair: 1000
    for(int j = 0; j < MAX_PRO_PAIR; j++){
      strcpy(comms[i].content[j], "");
    }
  }
  int comm_spe_count = 0;
  int prespe = -10000;
  FILE *f1, *f2;
  f1 = fopen("sorted_msa1", "r");
  f2 = fopen("sorted_msa2", "r");
  char buffer1[MAX_LIN_LENG],buffer2[MAX_LIN_LENG];
  while(fgets(buffer1, MAX_LIN_LENG, f1)){
    char msa1[MAX_SEQ_LENG], def1[MAX_DEF_LENG];
    int spe1;
    sscanf(buffer1, "%s  %d.%s", msa1, &spe1, def1);
    //printf("from ligmsa: %s %d.%s\n", msa1, spe1, def1);
    while(fgets(buffer2, MAX_LIN_LENG, f2)){
      char msa2[MAX_SEQ_LENG], def2[MAX_DEF_LENG];
      int spe2;
      sscanf(buffer2, "%s  %d.%s", msa2, &spe2, def2);
      //printf("from recmsa: %s %d.%s\n", msa2, spe2, def2);
      // a pair with new species
      if(spe1 == spe2 && spe1 != prespe){
        comms[comm_spe_count].count = 0;
        prespe = spe1;
        comms[comm_spe_count].spe = spe1;
	char temp[MAX_DEF_LENG];
	sprintf(temp, "%d.%s %d.%s", spe1, def1, spe1, def2);
	//printf("%s\n", temp);
	strcpy(comms[comm_spe_count].content[comms[comm_spe_count].count], temp);
	comms[comm_spe_count].count++;
	comm_spe_count++;
	if(comm_spe_count == MAX_SPE_PAIR){
	  break;
	}
      }
      // a pair with existed species
      else if(spe1 == spe2 && spe1 == prespe){
	char temp[MAX_DEF_LENG];
	sprintf(temp, "%d.%s %d.%s", spe1, def1, spe1, def2);
        //printf("%s\n", temp);
	strcpy(comms[comm_spe_count-1].content[comms[comm_spe_count-1].count], temp);
        comms[comm_spe_count-1].count++;
      }
    }
    fseek(f2, 0, SEEK_SET);
    if(comm_spe_count == MAX_SPE_PAIR){
      break;
    }
  }
  fclose(f1); fclose(f2);
  // show the comm species
  for(int index = 0; index < comm_spe_count; index++){
    printf("The following are common protein pairs for species: %d\n", comms[index].spe);
    for(int index2 = 0; index2 < comms[index].count; index2++){
      printf("%s\n", comms[index].content[index2]);
    }
  }
  printf("Done.\n");

  // find the real link and the link score if possible via the species based link files
  // this is the most time-consuming part of this program
  printf ("Build complex links from the common species ... \n");
  FILE *fout = fopen("cpx_link.txt", "w");
  for(int index = 0; index < comm_spe_count; index++){
    char newlinkfile[MAX_NAM_LENG];
    sprintf(newlinkfile, "%s.%d", linkfile, comms[index].spe);
    printf("searching specie pair %d from string links %s ...\n", comms[index].spe, newlinkfile);
    FILE *fin = fopen(newlinkfile, "r");
    char buffer[MAX_LIN_LENG];
    int startindex = 0;
    while(fgets(buffer, MAX_LIN_LENG, fin)){
      for(int index2 = startindex; index2 < comms[index].count; index2++){
        if(strstr(buffer, comms[index].content[index2]) != 0){
          fprintf(fout, "%s", buffer);
          // next time will start searching from the next protein pair, not from the beginning
          startindex = index2+1;
          break;
        }
      }
    }
    fclose(fin);
  }
  fclose(fout);
  printf("Done.\n");

  return 0;
}

// step 4: sort the real link result with sore ranked from high to low
// select the link with the highest score when there're several
// links between one protein with several different proteins
int sort_links(){
  printf("Sort complex links and select only the top 1 complex link for each protein ... \n");
  system("sort -k1,1 -k3nr,3  cpx_link.txt > cpx_link_sort13.txt");
  char initDef[MAX_DEF_LENG] = "XXX";
  char line[MAX_LIN_LENG];
  FILE *f1 = fopen("cpx_link_sort13.txt", "r");
  FILE *f2 = fopen("cpx_link_sort13_top1.txt", "w");
  while(fgets(line, MAX_LIN_LENG, f1)){
    char def1[MAX_DEF_LENG];
    sscanf(line, "%s ", def1);
    if(strcmp(def1, initDef) != 0){
      fprintf(f2, "%s", line);
      strcpy(initDef, def1);
    }
  }
  fclose(f1); fclose(f2);

  system("sort -k2,2 -k3nr,3 cpx_link_sort13_top1.txt > cpx_link_sort23.txt");
  strcpy(initDef, "XXX");
  f1 = fopen("cpx_link_sort23.txt", "r");
  f2 = fopen("top_cpx_link.txt", "w");
  while(fgets(line, MAX_LIN_LENG, f1)){
    char def1[MAX_DEF_LENG], def2[MAX_DEF_LENG];
    sscanf(line, "%s %s ", def1, def2);
    if(strcmp(def2, initDef) != 0){
      fprintf(f2, "%s", line);
      strcpy(initDef, def2);
    }
  }
  fclose(f1); fclose(f2);
  printf("Done.\n");
  return 0;
}


// build the MSA file for the whole protein complex 
// the default medium-high cutoff link score is 400 [400, 700]
// the high link score is [700, 900]
// the ultra high link score cutoff is [900, 999]
int build_cpx_msa(){
  printf("Build complex MSA for protein complex ... \n");
  int CUTOFF = 1;
  FILE *f1 = fopen("top_cpx_link.txt", "r");
  FILE *f2 = fopen(cpxoutfile, "w");
  char buffer[MAX_LIN_LENG];
  while(fgets(buffer, MAX_LIN_LENG, f1)){
    char def1[MAX_DEF_LENG], def2[MAX_DEF_LENG];
    int score;
    sscanf(buffer, "%s %s %d", def1, def2, &score);
    char msa1[MAX_LIN_LENG], msa2[MAX_LIN_LENG];

    FILE *f3 = fopen("msa1", "r");
    char line[MAX_LIN_LENG];
    while(fgets(line, MAX_LIN_LENG, f3)){
      char msa[MAX_LIN_LENG], def[MAX_DEF_LENG];
      sscanf(line, "%s  %s", msa, def);
      if(strcmp(def, def1) == 0){
        strcpy(msa1, msa);
      }
    }
    fclose(f3);

    f3 = fopen("msa2", "r");
    while(fgets(line, MAX_LIN_LENG, f3)){
      char msa[MAX_LIN_LENG], def[MAX_DEF_LENG];
      sscanf(line, "%s  %s", msa, def);
      if(strcmp(def, def2) == 0){
        strcpy(msa2, msa);
      }
    }
    fclose(f3);

    if(score > CUTOFF){
      fprintf(f2, "%s %s %.3f\n", msa1, msa2, (double)score/1000.0);
    }
  }
  fclose(f1); fclose(f2);
  printf("Done.\n");
  return 0;
}

int delete_files(){
  system("rm msa1 msa2 sorted_msa1 sorted_msa2");
  system("rm -rf cpx_link_sort13.txt");
  system("rm -rf cpx_link_sort13_top1.txt");
  system("rm -rf cpx_link_sort23.txt");
  system("rm -rf cpx_link.txt");
  system("rm -rf top_cpx_link.txt");
  return 0;
}

int main(int argc, char **argv){
  if(argc != 9){
    printf("------------------------------------ERROR-----------------------------------------------------------\n");
    printf("This program is used to build sequence alignment for a dimeric protein, the values in [] are required, ");
    printf("and the ligand and receptor msa file should be put in the current directory\n");
    printf("Usage: build_cpx_alignment -link [STRING_protein_links] -ligmsa [ligand.msa] -recmsa [receptor.msa] -out [protein_cpx_alignment.txt]\n");
    exit(-1);
  }
  
  linkfile = argv[2];
  msa1file = argv[4];
  msa2file = argv[6];
  cpxoutfile = argv[8];


  make_new_files();
  sort_msa();
  find_common_species_and_build_links();
  sort_links();
  build_cpx_msa();
  delete_files();

  return 0;
}

File: bin/ssip/seqalign/build_cpx_alignment.pl
#!/usr/bin/perl -w
###################################################################################################
# build_interface_alignment.pl
#
# This is a perl script used to build the protein complex MSA from two monomer MSAs via the STRING 
# protein links database.
# Input:
#         <-- protein.links.txt
#         <-- pdb_lig.msa
#         <-- pdb_rec.msa
# Output:
#         --> cpx_alignment.txt
#
# Xiaoqiang Huang
# xiaoqiah@umich.edu
# 08/01/2018
###################################################################################################

use 5.010;
use strict;
use warnings;
use Cwd 'abs_path';
use File::Basename;

if(@ARGV != 8){
  print "This script build_cpx_alignment.pl is used to build sequence alignment for a dimeric protein, the values in [] are required\n";
  print "Usage: build_cpx_alignment.pl -link [STRING_protein_links] -ligmsa [ligand.msa] -recmsa [receptor.msa] -out [protein_cpx_alignment.txt]\n";
  exit;
}

my $filepath = dirname(abs_path(__FILE__));

my $link = $ARGV[1];
my $msa1 = $ARGV[3];
my $msa2 = $ARGV[5];
my $cpxout = $ARGV[7];

# step 1: sort and read the multiple sequence alignment data
print "Sorting ligand and receptor MSA files ... ";
system "sort -k2 $msa1 > sorted_$msa1";
system "sort -k2 $msa2 > sorted_$msa2";
open MSA1, "<sorted_$msa1";
open MSA2, "<sorted_$msa2";
my @lines1 = <MSA1>;
my @lines2 = <MSA2>;
chomp(@lines1);
chomp(@lines2);
close MSA1;
close MSA2;
print "Done.\n";

# step 2: find common organisms of the two MSAs and build a pseudolink 
# between two proteins for each organism
print "Find common species of two MSAs and build links between proteins ... ";
my @comm = ();
for(my $k = 0; $k < @lines1; $k++){
  my $def1 = $lines1[$k];
  my @buff1 = split(/\s+/, $def1);
  $def1 = $buff1[1];
  my @buff2 = split(/\./,$def1);
  my $spe1 = $buff2[0];
  for(my $j = 0; $j < @lines2; $j++){
    my $def2 = $lines2[$j];
    my @buff3 = split(/\s+/, $def2);
    $def2 = $buff3[1];
    my @buff4 = split(/\./,$def2);
    my $spe2 = $buff4[0];
    if($spe1 == $spe2){
      push @comm, "$def1 $def2";
    }
  }
}
print "Done.\n";

# step 3: find the real link and the link score if possible via the species based link files
# this is the most time-consuming part of this program
print "Build complex links from the common species ... ";
my $firstLine = 1;
for(my $k = 0; $k < @comm; $k++){
  my $temp = $comm[$k];
  my @buff = split(/\./, $temp);
  my $speindex = $buff[0];
  my @buff2 = split(/\s+/, $temp);
  my $def1 = $buff2[0];
  if($firstLine == 1){
    `grep -r "$temp" $link.$speindex > cpx_link.out`;
    $firstLine = 0;
  }
  else{
    `grep -r "$temp" $link.$speindex >> cpx_link.out`;
  }
}
print "Done.\n";

# step 4: sort the real link result with sore ranked from high to low
# select the link with the highest score when there're several
# links between one protein with several different proteins
print "Sort complex links and select only the top 1 complex link for each protein ... ";
system "sort -k1,1 -k3nr,3  cpx_link.out > cpx_link_sort13.out";
open FILE, "<cpx_link_sort13.out";
my $initDef = "XXX";
$firstLine = 1;
while(my $line = <FILE>){
  chomp($line);
  my @buff = split(/\s+/, $line);
  my $def1 = $buff[0];
  if($def1 ne $initDef){
    if($firstLine == 1){
      `echo $line > cpx_link_sort13_top1.out`;
      $initDef = $def1;
      $firstLine = 0;
    }
    else{
      `echo $line >> cpx_link_sort13_top1.out`;
      $initDef = $def1;
    }
  }
}
close FILE;
system "sort -k2,2 -k3nr,3 cpx_link_sort13_top1.out > cpx_link_sort23.out";
open FILE, "<cpx_link_sort23.out";
$initDef = "XXX";
$firstLine = 1;
while(my $line = <FILE>){
  chomp($line);
  my @buff = split(/\s+/, $line);
  my $def1 = $buff[1];
  if($def1 ne $initDef){
    if($firstLine == 1){
      `echo $line > top_cpx_link.out`;
      $initDef = $def1;
      $firstLine = 0;
    }
    else{
      `echo $line >> top_cpx_link.out`;
      $initDef = $def1;
    }
  }
}
close FILE;
print "Done.\n";

# step 5: build the MSA file for the whole protein complex 
# the default medium-high cutoff link score is 0.4 [0.4, 0.7]
# the high link score is (0.7, 0.9]
# the ultra high link score cutoff is (0.9, 1.0)
print "Build complex MSA for protein complex ... ";
my $CUTOFF = 0.7;
open FILE, "<top_cpx_link.out";
open FILE2, ">$cpxout";
while(my $line = <FILE>){
  chomp($line);
  my @buff = split(/\s+/, $line);
  my $def1 = $buff[0];
  my $def2 = $buff[1];
  my $line1 = `grep -r $def1 $msa1`;
  my @temp1 = split(/\s+/, $line1);
  my $seq1 = $temp1[0];
  my $line2 = `grep -r $def2 $msa2`;
  my @temp2 = split(/\s+/, $line2);
  my $seq2 = $temp2[0];
  if($buff[2] > $CUTOFF){
    print FILE2 "$seq1 $seq2\n";
  }
}
close FILE;
close FILE2;
print "Done.\n";

system "rm sorted_$msa1 sorted_$msa2";
system "rm -rf cpx_link_sort13.out";
system "rm -rf cpx_link_sort13_top1.out";
system "rm -rf cpx_link_sort23.out";
system "rm -rf cpx_link.out";
system "rm -rf top_cpx_link.out";
